src/node/rpc/frontend.h

```diff
@@ -218,66 +218,25 @@ namespace ccf
 
       Store::Tx tx;
 
-      // Retrieve id of caller
-      std::optional<CallerId> caller_id;
-      if (ctx->is_create_request)
-      {
-        caller_id = INVALID_ID;
-      }
-      else
-      {
-        caller_id = handlers.valid_caller(tx, ctx->session->caller_cert);
-      }
-
-      if (!caller_id.has_value())
-      {
-        ctx->set_response_status(HTTP_STATUS_FORBIDDEN);
-        ctx->set_response_body(invalid_caller_error_message());
-        return ctx->serialise_response();
-      }
-
-      const auto signed_request = ctx->get_signed_request();
-      if (signed_request.has_value())
-      {
-        if (
-          !ctx->is_create_request &&
-          !verify_client_signature(
-            ctx->session->caller_cert,
-            caller_id.value(),
-            signed_request.value()))
-        {
-          set_response_unauthorized(ctx);
-          return ctx->serialise_response();
-        }
-
-        // Client signature is only recorded on the primary
-        if (
-          consensus == nullptr || consensus->is_primary() ||
-          ctx->is_create_request)
-        {
-          record_client_signature(
-            tx, caller_id.value(), signed_request.value());
-        }
-      }
+      auto caller_id = handlers.get_caller_id(tx, ctx->session->caller_cert);
 
       if (consensus != nullptr && consensus->type() == ConsensusType::PBFT)
       {
-        auto rep = process_if_local_node_rpc(ctx, tx, caller_id.value());
+        auto rep = process_if_local_node_rpc(ctx, tx, caller_id);
         if (rep.has_value())
         {
           return rep.value();
```

Comment:
Unrelated to this PR, but while you're here: This should just be `return rep;`, because we don't need to de-opt and re-opt, we can just return this directly, right? Except apparently `process_if_local_node_rpc` still returns `std::optional<nlohmann::json>`! Oops! That should be `std::optional<std::vector<uint8_t>>` like everything else, and we shouldn't be converting it to and from a JSON array.

Reply:
Good point! `forward_or_redirect_json()` also returned a `std::optional<nlohmann::json>` so I've fixed it too.